### DISCUSSION TITLE
Creating a `:syntax` item for all files regardless of filetype

### DISCUSSION THREAD

**Discussion by syyyr** [neovim/neovim#26902] [Q&A] [answered]
Hi, I'd like to have git merge conflicts highlighted in neovim at all times.
```vim
syn match MergeConflict '\v^[<\|=|>]{7}([^=].+)?$'
```
I want them to be highlighted all the time, for any filetype, even with no filetype.
So what I do is I put this autocmd into my init.lua:
```lua
vim.api.nvim_create_autocmd('Syntax', {
    callback = function()
        vim.cmd([[syn match MergeConflict '\v^[<\|=|>]{7}([^=].+)?$']])
    end,
    group = vim.api.nvim_create_augroup('CustomSyntax', {clear = true})
})
```
However, there's a bit of a problem: the procedure for syntax enabling runs synload.vim and that file sets up an autocmd that removes all syntax items (via `syn clear` - [here](https://github.com/neovim/neovim/blob/master/runtime/syntax/synload.vim#L29)), which means that my syntax item is immediately removed, because my autocmd is set up earlier, than the one in synload.vim. I have a few solutions, but none of them seem correct/right:
1) Change the autocmd to something like `BufEnter` - works, but runs everytime I switch buffers.
2) Provide my own synload.vim, which calls the original synload.vim, and then sets up my autocmds for syntax - works, but seems more like a hack, especially because I have to source the file by full path (`source /usr/share/nvim/runtime/syntax/synload.vim`)
3) Call `syntax enable` manually, so that I can define my autocmds after the ones in synload.vim - works, but syntax is enabled in neovim by default , so I'm not sure if I want to enable it manually like this. Also, it causes an error if called before setting up treesitter parsers via nvim-treesitter (might only be the ones that should be included with the installation (like lua) - I have none included, because I use nightly versions). 
4) Set my custom syntax somewhere else / use different autocmd-event - I'm not sure where, or what autocmd-event I should use... maybe I'm missing something.

Also, of these solution don't work if a file has no filetype unfortunately. I'm not sure how to set the syntax there. Possibly by the manual call to `syntax enable`?

**Comment by clason** [comment]
Do not use `syntax` and just `:match` and put that, e.g., in a `WinEnter` autocommand.

**Reply by syyyr** [reply]
With this init.lua
```lua
vim.cmd('highlight MergeConflict ctermbg=black ctermfg=red guibg=black guifg=red')
vim.api.nvim_create_autocmd('WinEnter', {
    callback = function()
        vim.cmd([[match MergeConflict /\v^[<\|=|>]{7}([^=].+)?$/]])
    end,
    group = vim.api.nvim_create_augroup('MySyntax', {clear = true})
})
```
it does kinda work, but maybe I have to use a different autocmd? The highlighting doesn't appear until I at least split the window. And then the highlighting only appears in the second split after I enter it (which is obvious, since I used the `WinEnter` event).

**Reply by clason** [reply]
Try adding `WinNew` to the events.

**Reply by syyyr** [reply]
I understand that with a sufficient amount of events, I'll maybe eventually get it, but I was really wondering about whether there was I way to add/override syntax in a less fragile way. For example, it is very easy to change syntax of a single filetype - just add some lines to after/syntax/\<filetype\>.lua. So I was looking for something like that, but regardless of filetype.

Best I can do now is this:
```lua
vim.api.nvim_create_autocmd('BufEnter', {
    callback = function()
        if vim.b.custom_syntax_set then
            return
        end
        vim.b.custom_syntax_set = true
        vim.cmd([[syn match MergeConflict '\v^[<\|=|>]{7}([^=].+)?$']])
    end,
    group = vim.api.nvim_create_augroup('CustomSyntax', {clear = true})
})
```
Not very clean, but it works I guess, and it seems like the best way I found so far.

**Reply by clason** [reply, marked as answer]
There is no "global syntax", no. `:match` is exactly meant for syntax-independent highlighting like you are looking for. The exact events depend a bit on your workflow; note that `:match` is window-local, though, so `BufWinEnter` may be better.

You might also want to look at the [mini.hipatterns](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-hipatterns.md) plugin; either for using it directly or as a basis of your own implementation.

**Reply by syyyr** [reply]
I understand. Thank you very much for the help.

This is what I have for now:
```lua
vim.api.nvim_create_autocmd({'BufWinEnter', 'WinEnter'}, {
    callback = function()
        if vim.w.custom_syntax_set then
            return
        end
        vim.w.custom_syntax_set = true
        vim.cmd([[match MergeConflict /\v^[<\|=|>]{7}([^=].+)?$/]])
    end,
    group = vim.api.nvim_create_augroup('CustomSyntax', {clear = true})
})
```
Later, I'll try looking into mini.hipatterns.
